### PR TITLE
Introduce StagePath for working with files on stages

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
1. Renames existing `StagePath` to `StagePathType` to avoid ambiguity
2. Introduce `StagePath` class representing a file on stage. This class offers `Path` like operations including `joinpath` or `parent` operations.

The next step of refactor will be removing of `StagePathParts` concepts and then moving query generation to `StagePath`.

### CODE REVIEW

File: tests/git/test_git_commands.py [hunk 1]
```diff
@@ -542,35 +544,35 @@ def test_api_integration_and_secrets_get_unique_names(
...
         (
             "@repo/branches/main/",
-            "@repo/branches/main/",
+            "@repo/branches/main",
             ["@repo/branches/main/s1.sql", "@repo/branches/main/a/S3.sql"],
         ),
         (
             "@repo/branches/main/a",
-            "@repo/branches/main/",
+            "@repo/branches/main",
```
Comment (sfc-gh-turbaszek) [inline tests/git/test_git_commands.py:552]: To align the behavior. The `StagePath` accepts both inputs (with or without trailing /) but the representation is always without / suffix

**Comment by sfc-gh-turbaszek** [comment]
Skipping the one of NA test because is failing also on other PRs
https://github.com/snowflakedb/snowflake-cli/pull/1691

File: src/snowflake/cli/_plugins/stage/diff.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 
 log = logging.getLogger(__name__)
 
-StagePath = PurePosixPath  # alias PurePosixPath as StagePath for clarity
+StagePathType = PurePosixPath  # alias PurePosixPath as StagePath for clarity
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/_plugins/stage/diff.py:33]: This change is done to avoid ambiguity

**Review by sfc-gh-astus — APPROVED** [review]

**Review by sfc-gh-fcampbell — APPROVED** [review]
NADE LGTM

**Pull request merged**